pyfifotax/report_data.py

```diff
@@ -0,0 +1,462 @@
+                    currency=event.currency,
+                    date=event.date,
+                    value=event.bonus,
+                )
+
+                bought_shares = awv.AWVEntryZ10ESPPDeposit(
+                    date=event.date,
+                    symbol=event.symbol,
+                    currency=event.currency,
+                    quantity=event.received_shares.quantity,
+                    value=event.received_shares.total_buy_value(),
+                )
+
+                self.awv_z4_events.append(bonus)
+                self.awv_z10_events.append(bought_shares)
+
+            elif isinstance(event, DividendEvent):
+                self.held_forex[event.currency].push(event.received_net_dividend)
+
+                if event.withheld_tax is not None:
+                    self.misc["Tax Withholding"].append(event.withheld_tax)
+                self.misc["Dividend Payments"].append(event.received_dividend)
+
+                # dividends should be small enough to not trigger AWV reportings
+
+            elif isinstance(event, BuyEvent):
+                # if not enough money, pop on FOREX Queue will fail
+                tmp = self.held_forex[event.currency].pop(
+                    event.cost_of_shares,
+                    to_decimal(1),
+                    event.date,
+                )
+                self.sold_forex[event.currency].extend(tmp)
+
+                self.held_shares[event.symbol].push(event.received_shares)
+                if event.paid_fees is not None:
+                    self.misc["Fees"].append(event.paid_fees)
+
+                self.awv_z10_events.append(
+                    awv.AWVEntryZ10Buy(
+                        date=event.date,
+                        symbol=event.received_shares.symbol,
+                        currency=event.currency,
+                        quantity=event.received_shares.quantity,
+                        value=event.received_shares.total_buy_value(),
+                    )
+                )
+
+            elif isinstance(event, SellEvent):
+                # if not enough shares to sell, pop on SHARE Queue will fail
+                tmp = self.held_shares[event.symbol].pop(
+                    event.quantity, event.sell_price, event.date
+                )
+                self.sold_shares[event.symbol].extend(tmp)
+
+                self.held_forex[event.currency].push(event.received_forex)
+
+                if event.paid_fees is not None:
+                    self.misc["Fees"].append(event.paid_fees)
+
+                self.awv_z10_events.append(
+                    awv.AWVEntryZ10Sale(
+                        date=event.date,
+                        symbol=event.symbol,
+                        currency=event.currency,
+                        quantity=event.quantity,
+                        value=event.quantity * event.sell_price,
+                    )
+                )
+
+            elif isinstance(event, CurrencyConversionEvent):
```

Comment:
On the same day currency conversion events should take priority before buy events or it can happen when the person first converts the currency then buys something with it on the same day, the buy event will throw an exception.
Reproducer: [forex_first.xlsx](https://github.com/user-attachments/files/15920198/forex_first.xlsx)

Similarly, sell others first on the same day to obtain the currency to buy another equity with it.
Reproducer: [sell_others_first.xlsx](https://github.com/user-attachments/files/15920232/sell_others_first.xlsx)

Reply:
Just basing things on dates is the main culprit. People could buy things by accident and directly sell in which buy should take precedence. 
The second point couldn't happen with the main focus on Schwab thanks to T+2/T+1. 
I personally don't really know what the best way forward is.
Deposits are clearly first, splits are clearly last. 
Sell, then Currency Conversions, then Buy should cover your edges cases, right? The only thing not covered would then be buy and sell directly after.

Reply:
I definitely see now why T+1 makes sense. You avoid a lot of these issues if you have settlement periods :)
Doing the balance only EOD definitely makes sense. However, this definitely complicates creating the FIFO matching.

Reply:
- I'll fix the first case by splitting conversions into EUR->FOREX and FOREX->EUR, we don't really care about the EUR balance too much.
- the second case can be fixed by swapping buy/sell priorities, however, this now rules out directly selling stocks at the same day they have been bought - on the other hand, this case should be less likely than using proceeds to buy something else